お部屋のスコア入力ページを作成する

uchihiro04 opened the pull request:
## 概要 
お部屋のスコア入力ページを表示しました。
また、「スコアを保存する」ボタンを押すと、入力したスコアが保存されお部屋の内見一覧ページにリダイレクトするように設定しました。

## スクリーンショット
* お部屋のスコア入力ページ
<img width="700" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/24bb7b8b-0158-4230-bb84-ea734ac7289f">

* デベロッパツールで「iPhone SE」の画面（375 × 667）を想定して表示した場合
<img width="200" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/3dc67bd7-92df-4058-a65b-4ec46ea7568d">

* 作成後の画面
お部屋の内見一覧ページにリダイレクトし、「スコアを登録しました」と表示されます。
<img width="700" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/a7be8039-5faa-4e95-b5fe-ba2ad777a49a">

デベロッパツールで「iPhone SE」の画面（375 × 667）を想定して表示した場合
<img width="200" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/e3f03d7c-6ad5-4e37-ae03-ab5f9e0beda4">

## 関連Issue
- #107 

## 備考
以下の実装は別Issueで行います。
* 入力者一覧の作成
* 日本語化対応
* デザイン

Close #107

uchihiro04 committed:
#107 コントローラのひな形作成

HouseViewing::Rooms::Scoresコントローラのひな形を作成しました。

uchihiro04 committed:
#107 ルーティングを設定

house_viewings/rooms/scores#newコントローラに関するPathを設定しました。

uchihiro04 committed:
#107 見出しを作成

スコア入力ページについて、見出しを作成しました。

uchihiro04 committed:
#107 部屋名を表示

スコアをつける部屋名を表示しました。部屋名に関するフォームには、初期値として現時点での部屋名を入れています。

uchihiro04 committed:
#107 入力者を表示

入力者の項目を表示しました。現時点ではラベルの日本語化は対応していません。

uchihiro04 committed:
#107 Slim-Lintの指摘点を修正

Slim-Lintの指摘を受け、ラベル名や値を修正しました。

uchihiro04 committed:
#107 入力者に関する見出しを修正

入力者について、labelタグからh2タグに変更しました。フォームへのタグではなく、ページ全体の見出しとして表現する方が適切なのではないかと考えたためです。

uchihiro04 committed:
#107 フォーム内の項目を共通化

スコアの編集ページと同じ部品になるため、フォーム内の項目を共通化しました。

uchihiro04 committed:
#107 評価項目に関するコードの削減

評価項目に関するコードがDRYでなかったため、ヘルパーメソッドなどを用いてコードの削減を行いました。

uchihiro04 committed:
#107 localsオプションの修正

renderのlocalsオプションについて、Slim-Lintからの指摘点を修正しました。

uchihiro04 committed:
#107 「スコアを保存する」ボタンの表示

「スコアを保存する」ボタンを表示しました。現時点ではボタンを押してもデータは保存されません。

uchihiro04 committed:
#107 createアクションを作成

uchihiro04 committed:
#107 フォームの送信先を修正

フォームの送信先がnewアクションに関するPathになっていたため、createアクションに関するPathに修正しました。

uchihiro04 committed:
#107 フラッシュメッセージが表示されるよう設定

uchihiro04 committed:
#107 戻るボタン（<）の作成

戻るボタン(<)を作成しました。ボタンを押すと「お部屋の内見一覧ページ」に遷移します。

uchihiro04 committed:
#107 存在しないroom_idでアクセスできる不具合を修正

スコア入力ページについて、存在しないroom_idでもアクセスできてしまう不具合を修正しました。

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/_form.html.slim:
```diff
@@ -0,0 +1,23 @@
+          = specific_bad_evaluation_name(evaluation_item)
+      - else
+        p 良い
+        p 悪い
```
```
- if specific_evaluation_names?(evaluation_item)
        p
          = specific_good_evaluation_name(evaluation_item)
        p
          = specific_bad_evaluation_name(evaluation_item)
      - else
        p 良い
        p 悪い
```
ここは↓みたいな感じがいいですかね

view 
```
p = good_evaluation_name(evaluation_item)
p = bad_evaluation_name(evaluation_item)
```

helper
```
def good_evaluation_name(evaluation_item)
SPECIFIC_EVALUATION_NAMES.dig(evaluation_item.to_sym, :good) | '良い'
end

def bad_evaluation_name(evaluation_item)
p = SPECIFIC_EVALUATION_NAMES.dig(evaluation_item.to_sym, :bad) | '悪い'
end
```

uchihiro04 replied:
レビューいただいた方法だと、条件分岐を使わなくて済むためコードも簡潔になりますね……。
いただいたコメントをもとに、条件分岐に関するコードを削除しました。

good_evaluation_nameメソッドおよびbad_evaluation_nameメソッドで`|`を使用したところ、以下のエラーが発生しました。そのため、`|`の代わりに`||`を使用しました。
```
ActionView::Template::Error (undefined method `|' for "不足":String):
    .flex
      p = good_evaluation_name(evaluation_item)
      p = bad_evaluation_name(evaluation_item)
```

該当するコミットは[c129391](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/c1293911f57c7578e2925c9a46feceef25d86a42)です。

uchihiro04 committed:
#107 評価基準に関する条件分岐を削除

ヘルパーメソッドを修正することで、スコア入力ページの評価基準に関する条件分岐に関する部分を削除しました。

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/new.html.slim:
```diff
@@ -0,0 +1,4 @@
+h1
+  = "#{@room.name}を入力する"
+= form_with url: new_house_viewing_room_score_path, local: true do |form|
```
urlではなくroom modelを渡してあげるのでいいと思います
roomとscoreのformを分けるために`|form|`を`|room_form|`にしちゃいましょう

uchihiro04 commented:
Slim-Lintについて相談させていただければと思います。

【内容】
Slim-LintのLineLength（1行の文字数）について、制限を80字から100字もしくは120字に変更してもよろしいでしょうか。

【経緯】
今回のアプリではPathが長くなっています。そのため、`link_to`などでPathを記載した場合Slim-Lintに引っかかる可能性が高そうです。
Pathの長さを考慮して制限を緩和して問題ないか確認するために、相談させていただけばと思います。

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/new.html.slim:
```diff
@@ -0,0 +1,4 @@
+h1
+  = "#{@room.name}を入力する"
+= form_with url: new_house_viewing_room_score_path, local: true do |form|
```
ここってurl: new_xxxではうまく動作しないはずなので、model: @roomにしましょう

レコードの有無でcreateかupdateかよしなに判別してくれます
参考: https://pikawaka.com/rails/form_with#form_with%E3%81%AE%E5%BC%95%E6%95%B0

そうするとこの部分もpartialで共通化できるはずです

また、roomとscoreのformを分けるため
|form|→|room_form|にちゃいましょう

takuminasu replied:
urlもきちんと設定すれば動きますが今回はmodel指定でも動くと思うのでmodelの指定がいいと思います。
以下のようなイメージです。
```
= form_with model: [@house_viewing, @room, @score]
```

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/_form.html.slim:
```diff
@@ -0,0 +1,17 @@
+= form.fields_for :room do |room_fields|
+  = room_fields.text_field :name, value: @room.name
```
```
= form.fields_for :room do |room_fields|
  = room_fields.text_field :name, value: @room.name
```

ではなく
`= room_form.text_field :name, value: @room.name`
でfields_for使わなくて大丈夫です

参考: https://railsguides.jp/form_helpers.html#fields-for%E3%83%98%E3%83%AB%E3%83%91%E3%83%BC

Yuma-Nagaoka commented:
> Slim-Lintについて相談させていただければと思います。
> 
> 【内容】 Slim-LintのLineLength（1行の文字数）について、制限を80字から100字もしくは120字に変更してもよろしいでしょうか。
> 
> 【経緯】 今回のアプリではPathが長くなっています。そのため、`link_to`などでPathを記載した場合Slim-Lintに引っかかる可能性が高そうです。 Pathの長さを考慮して制限を緩和して問題ないか確認するために、相談させていただけばと思います。

具体的にどこのコードが引っかかりましたか？

uchihiro04 commented:
>具体的にどこのコードが引っかかりましたか？

まだ実装してはいないのですが、「お部屋の内見一覧ページ」から各部屋に遷移する際のリンクを設定しようと考えていて、以下のようなコードを想定しました。
`= link_to room.name, new_house_viewing_room_score_path(room_id: room.id)`

このコードだとSlim-LintのLineLengthに引っかかるため、今後同じようなことが何度か起こるのではないかと思い相談した形です。

Yuma-Nagaoka commented:
> このコードだとSlim-LintのLineLengthに引っかかるため、今後同じようなことが何度か起こるのではないかと思い相談した形です。

なるほどです、120にしちゃっていいと思います。

uchihiro04 committed:
#107 Slim-Lintの設定を変更

Pathが含まれたコードがLintに引っかかることを考慮し、Slim-LintについてLineLengthの上限を80字から120字に変更しました。

uchihiro04 commented:
>なるほどです、120にしちゃっていいと思います。

ありがとうございます。SLim-LintのLineLengthについて、上限を80字から120字に変更しました。該当するコミットは[3ca5403](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/3ca540366226a0c8588a56039064791a9369b6dc)です。

uchihiro04 committed:
#107 お部屋の内見一覧ページにリンクを設定

お部屋の内見一覧ページについて、部屋名をクリックするとスコア入力ページに遷移するようにリンクを設定しました。

uchihiro04 committed:
#107 入力フォームを修正

スコア入力ページについて、form_withのオプションをurlからmodelに変更しました。また、1つのボタンでRoomモデルの更新とScoreモデルの新規作成ができるようにしました。

uchihiro04 committed:
#107 プライベートメソッドの設定

HouseViewingモデルについて、外部で使用しないメソッドについてプライベートメソッドを設定しました。

uchihiro04 committed:
#107 バリデーション失敗時のレイアウト崩れを修正

バリデーションに失敗した時にラジオボタンなどのレイアウトが崩れてしまう不具合を修正しました。

uchihiro04 committed:
#107 rubocopの指摘点を修正

Style/LineLengthやMetrics::AbcSizeについてrubocopから指摘を受けたため、before_actionの設定や変数の追加などを行い対応しました。

uchihiro04 committed:
内見一覧ページのお部屋の並び順を設定

内見一覧ページのお部屋の並び順について、お部屋の作成日順で設定しました。

takuminasu commented on app/controllers/house_viewings/rooms/scores_controller.rb:
```diff
@@ -0,0 +1,48 @@
+
+      def create
+        room_name = params[:score][:room][:name]
+        @room.update!(name: room_name) if @room.name != room_name
+        @score = @room.scores.new(score_params)
+
+        if @score.save
```
こちらtransactionをはるようにしましょう。

uchihiro04 replied:
transactionについて、1点ご相談させていただければと思います。

【内容】
Roomモデルの部屋名を空欄の状態で「スコアを保存する」を押すと、form_withでエラーが発生します。

【経緯】
Scoreコントローラについて、以下のようにトランザクションを追加しました。
```
      def create
        room_name = params[:score][:room][:name]
        Score.transaction do
          @room.update!(name: room_name) if @room.name != room_name
          @score = @room.scores.new(score_params)
          @score.save!
          redirect_to house_viewing_rooms_path, notice: 'スコアを登録しました。'
        end
      rescue ActiveRecord::RecordInvalid
        render :new, status: :unprocessable_entity
      end
```

その際、Roomモデルの部屋名を空欄にして「スコアを保存する」を押すと、以下のエラーが発生しました。
<img width="700" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/f070e292-8456-428d-af37-c02398e88bed">

エラー時は`@score`がnilになっているため、`@score`の値がエラーの原因かと推測しています。

Roomモデルの部屋名を空欄にして「スコアを保存する」を押してもエラーが発生しない方法が分からず、現在詰まっているため、ご相談させていただければと思います。

uchihiro04 replied:
レビューおよびハドルでのアドバイスをもとに、以下のような形で変更を行いました。
* transactionの追加
* HouseViewing::Rooms::Scoresコントローラのcreateメソッドについて、`@score`を`@room`の上に変更

該当するコミットは[f220142](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/f220142fe2ce7c57fdab0cbf2ea285c2f526ae45)です。

takuminasu commented:
エラー時にメッセージが表示されていないので表示されるようにお願いします。

https://github.com/uchihiro04/house-viewing-score-log/assets/62975075/6adee4d4-a0ce-4be1-b4e4-c8ebb29e76a3

uchihiro04 committed:
#107 pry-railsをインストール

デバッグ作業を円滑にするために、pry-railsをインストールしました。

uchihiro04 committed:
#107 transactionを追加

roomの更新とscoreの新規作成・保存が中途半端な状態にならないように、transactionを追加しました。

uchihiro04 committed:
#107 エラーメッセージの追加

バリデーションエラー時のエラーメッセージを追加しました。

uchihiro04 committed:
#107 スコア入力ページの見出しを変更

uchihiro04 commented:
>エラー時にメッセージが表示されていないので表示されるようにお願いします。

バリデーション失敗時のエラーメッセージを追加しました。該当するコミットは[8d42297](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/8d42297334ad1e26b58f06d15f629b6656491021)です。

---
エラーメッセージについて、1点ご相談です。

【内容】
ユーザーに見せるエラーメッセージについて、「カラム名 is not a number」を表示する必要はありますでしょうか。
また、表示しない場合はどういった方法を取るといいでしょうか。

【経緯】
スコア入力ページについて、バリデーション失敗時のエラーメッセージは以下の通りです。
<img width="200" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/35b2df57-7f5d-43f2-8d15-bef7c992c12d">

Scoreの評価項目では1から5の数値しか入力できないバリデーションを設定しています。そのため、評価項目が未入力だと「カラム名 is not a number」というエラーメッセージが表示されます。
しかし、ユーザーからしたら「カラム名 can't be blank」が表示されていれば問題ないように思います。

そのため、いずれかの方法で「カラム名 is not a number」をエラーメッセージで表示しない形を考えています。
* エラーメッセージの配列から「カラム名 is not a number」を除外して表示する
* 該当するバリデーションを削除する

個人的には、通常評価項目について1−5以外の数値を入力する方法がないため、バリデーション自体を削除しても良いのかなと考えています。

今回判断に迷いましたため、ご相談させていただければと思います。

uchihiro04 committed:
#107 フラッシュメッセージの条件分岐を削除

Slim-Lintの指摘を受け、フラッシュメッセージの条件分岐の代わりにflash.notice.presenceを使用しました。

takuminasu commented on app/controllers/house_viewings/rooms/scores_controller.rb:
```diff
@@ -0,0 +1,48 @@
+        @score = Score.new
+      end
+
+      def create
```
トランザクションを使用せずに以下のような形でscoreのインスタンスを作成すればroomと同じタイミングで保存されました。
```ruby
def create
  room_name = params[:score][:room][:name]
  @score = @room.scores.new(score_params)
  @room.name = room_name if @room.name != room_name
    
  if @room.save
    redirect_to house_viewing_rooms_path, notice: 'スコアを登録しました。'
  else
    render :new, status: :unprocessable_entity
  end
end
```

uchihiro04 replied:
いただいたレビューをもとに、トランザクションを用いない形に変更しました。該当するコミットは[c927f48](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/c927f482ef9d2a7bd435716eef26ca2858b909ee)となります。

---
最初、accepts_nested_attributes_forなどを用いないでなぜ同時に保存できているのか分からなかったのですが、Railsガイドを確認して分かってきたため、解釈を以下に記載します。

https://railsguides.jp/association_basics.html#has-many%E9%96%A2%E9%80%A3%E4%BB%98%E3%81%91%E3%81%AE%E8%A9%B3%E7%B4%B0%E3%81%AA%E5%8F%82%E7%85%A7-%E3%82%AA%E3%83%96%E3%82%B8%E3%82%A7%E3%82%AF%E3%83%88%E3%81%8C%E4%BF%9D%E5%AD%98%E3%81%95%E3%82%8C%E3%82%8B%E3%82%BF%E3%82%A4%E3%83%9F%E3%83%B3%E3%82%B0 
 
上記リンク内には以下のような文章がありました。
>親オブジェクト（has_many関連付けを宣言している側のオブジェクト）が保存されない場合（つまりnew_record?がtrueを返す場合）、子オブジェクトは追加時に保存されません。親オブジェクトが保存されると、関連付けられていたオブジェクトのうち保存されていなかったメンバーはすべて保存されます。

上記の通り、親オブジェクト（`@room`）が保存されると、関連付けられていたオブジェクトのうち保存されていなかったもの（`@score`）もあわせて保存されます。そのため、いただいたコードで同時保存できると解釈しました。 

もし解釈が間違っていましたら、ご指摘いただけますと助かります。

takuminasu commented on app/views/house_viewings/rooms/scores/new.html.slim:
```diff
@@ -0,0 +1,5 @@
+  = link_to '<', house_viewing_rooms_path
+  h1
+    = 'スコアを入力する'
+= render partial: 'form'
```
localsオプションを利用して共通テンプレートではインスタンス変数を使用しないようにしましょう。

uchihiro04 replied:
確かに直接インスタンス変数を用いると、localsオプションを用いた時と比べてコントローラと共通テンプレートの結合度が高くなりますね……。

レビューをもとにlocalsオプションを使用する形に変更しました。該当するコミットは[5b9a659](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/5b9a659342285d7a8e9ae4bd2226997c202e1800)です。

takuminasu commented:
>そのため、いずれかの方法で「カラム名 is not a number」をエラーメッセージで表示しない形を考えています。

エラーメッセージの制御については一旦このPRでは考慮したくないです。

uchihiro04 committed:
#107 roomレコード保存時にscoreレコードを保存する形に変更

createアクションについて、transactionの使用をやめてroomレコード保存時にscoreレコードも保存する形に変更しました。

uchihiro04 committed:
#107 rubocopの指摘点を修正

不要なスペースが空いていたため、修正を行いました。

uchihiro04 committed:
#107 renderにlocalsオプションを設定

レビューでの指摘を受け、renderにlocalsオプションを設定し、ビューが直接インスタンス変数を使用しない形に変更しました。

takuminasu commented on app/views/house_viewings/rooms/scores/_form.html.slim:
```diff
@@ -0,0 +1,25 @@
+      = form.label evaluation_item.to_sym, i + 1
+      = form.radio_button evaluation_item.to_sym, i + 1
+  .block
+    = button_tag 'スコアを保存する', type: 'submit'
```
form.submitを利用しましょう。
https://railsdoc.com/page/submit

uchihiro04 replied:
button_tagからform.submitを使用した形に修正を行いました。該当するコミットは[3ffabe9](https://github.com/uchihiro04/house-viewing-score-log/pull/112/commits/3ffabe9da9c67011b5ec541910a121948529850b)です。

uchihiro04 committed:
#107 フォームの送信に関するメソッドを変更

フォームの送信で使用するメソッドについて、button_tagからsubmitに変更しました。

takuminasu reviewed (approved):
LGTM
